url_matcher/patterns.py

```diff
@@ -0,0 +1,265 @@
+    if not has_scheme:
+        pattern = f"//{pattern}"
+    elif pattern.startswith("//"):
+        # This is required because urlparse("//example.com").netloc == "//example.com"
+        # but instead we want it to be parsed into the the path. We achieve it by appending
+        # two more slashes
+        pattern = f"//{pattern}"
+    return pattern
+
+
+ParseTuple = namedtuple("ParseTuple", "scheme netloc path query fragment")
+
+
+@lru_cache(30)
+def pattern_parse(pattern: str) -> ParseTuple:
+    """
+    Parses the pattern to a named tuple (scheme, netloc, path, query, fragment)
+    >>> pattern_parse("example.com")
+    ParseTuple(scheme='', netloc='example.com', path='', query='', fragment='')
+    >>> pattern_parse("//example.com/path;this_is_also_path")
+    ParseTuple(scheme='', netloc='', path='//example.com/path;this_is_also_path', query='', fragment='')
+    """
+    pattern = pattern_to_url(pattern)
+    return _urlparse(pattern)
+
+
+def _urlparse(url: str) -> ParseTuple:
+    """
+    Returns a named tuple (scheme, netloc, path, query, fragment)
+    where path and params are joined together into path and
+    some other elements are normalized.
+
+    >>> _urlparse("scheme://example.com/path;params?query=23#fragment")
+    ParseTuple(scheme='scheme', netloc='example.com', path='/path;params', query='query=23', fragment='fragment')
+    >>> _urlparse("http://example.com:80/path")
+    ParseTuple(scheme='http', netloc='example.com', path='/path', query='', fragment='')
+    """
+    scheme, netloc, path, params, query, fragment = urlparse(url)
+    path = _join_path_and_params(path, params)
+    scheme, netloc = normalize_netloc_and_schema(scheme, netloc)
+    return ParseTuple(scheme, netloc, path, query, fragment)
+
+
+def _wildcard_re_escape(text: str):
+    return re.escape(text).replace("\\*", ".*")
+
+
+def _join_path_and_params(path, params):
+    if params:
+        return f"{path};{params}"
+    else:
+        return path
+
+
+def normalize_netloc_and_schema(schema: str, netloc: str) -> Tuple[str, str]:
+    """
+    Removes 80 or 443 port when obvious. Deduces http or https when the port is provided
+
+    >>> normalize_netloc_and_schema("http", "example.com:80")
+    ('http', 'example.com')
+    >>> normalize_netloc_and_schema("http", "example.com:80")
+    ('http', 'example.com')
+    >>> normalize_netloc_and_schema("http", "example.com:443")
+    ('http', 'example.com:443')
+    >>> normalize_netloc_and_schema("https", "example.com:443")
+    ('https', 'example.com')
+    >>> normalize_netloc_and_schema("", "example.com:80")
+    ('http', 'example.com')
+    >>> normalize_netloc_and_schema("", "example.com:443")
+    ('https', 'example.com')
+    """
+    schema = schema.lower()
+    domain, port = split_domain_port(netloc)
+    if (port == "80" and schema in ("http", "")) or (port == "443" and schema in ("https", "")):
+        return "http" if port == "80" else "https", domain
+    else:
```

Comment:
**_optional_**: The `else` statement could be dropped.

Reply:
Same answer as in https://github.com/zytedata/url-matcher/pull/2#discussion_r756838358